### PR TITLE
Mercator removal

### LINKED ISSUE #326: Remove scala-reflect from compile-time dependencies
It is pulled in transitively via mercator. Instead, it should be marked as provided.

Maybe it would be possible to remove the mercator dependency alltogether?

### PR DESCRIPTION
**Pull request by Pask423**
Should close #326

### CODE REVIEW

**Comment by lukaszlenart** [comment]
The build fails on _binary incompatibility_ check but I think that's ok.

File: core/src/main/scala/magnolia1/monadic.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  type Apply[X] = F[X]
+  def point[A](value: A): F[A]
+  def map[A, B](from: F[A])(fn: A => B): F[B]
+}
+
+trait Monadic[F[_]] extends Functor[F] {
+  def flatMap[A, B](from: F[A])(fn: A => F[B]): F[B]
+}
+
+object Monadic {
```
Comment (adamw) [inline core/src/main/scala/magnolia1/monadic.scala:13]: I assume this is copied from mercator, we should add a comment pointing back to the original source

File: core/src/main/scala/magnolia1/monadic.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
+package magnolia1
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Success, Try}
+
+trait Functor[F[_]] {
```
Comment (adamw) [inline core/src/main/scala/magnolia1/monadic.scala]: do we need the `Functor` as a stand-alone abstraction anywhere? My worry would be that it's such a common name, that we would add to the import-pollution when magnolia is on the classpath. Maybe we could simply have only `Monadic` with all the methods?

File: core/src/main/scala/magnolia1/monadic.scala (hunk 1 above)
Comment (Pask423) [reply core/src/main/scala/magnolia1/monadic.scala:13]: As I wrote in commit msg it was copied from scala3 branch so it depends on from where it was copied there

Comment (adamw) [reply core/src/main/scala/magnolia1/monadic.scala:13]: Ah :D I didn't read the commit msg, that's fine then

**Comment by joroKr21** [comment]
I would advise to not expose the instances as objects / classes in the interest of binary compatibility for the future.

**Comment by adamw** [comment]
@joroKr21 you mean expose them as `def`s instead? I didn't know that this has implications for bin compat, but I guess it makes sense, as a `def` gives you much more flexiblity

**Comment by joroKr21** [comment]
> @joroKr21 you mean expose them as defs instead? I didn't know that this has implications for bin compat, but I guess it makes sense, as a def gives you much more flexiblity

Yes, exactly - well if the object/class is public you can't remove it in a binary compatible way. And the would also be reflected in the types of the instances so you wouldn't be able to change that either.

**Comment by adamw** [comment]
@joroKr21 Thanks, makes sense :) We have to do the same change on the scala3 branch as well then

**Comment by Pask423** [comment]
Ok I will do it for scala2 but I think we need separate issue for scala3

**Comment by adamw** [comment]
@Pask423 yeah a separate PR for sure :)

**Pull request merged**